feat(annotations): add sdk for annotations api (CXT-488)

cyriaquebrousse opened the pull request:
Add basic support for Annotations API (see [specification](https://docs.cognite.com/api/playground/#tag/Annotations)).

Support for annotation types is not included in this PR for the sake of speed, which means that `annotationType` (str) is a hardcoded enum in this PR, and `data` is a generic object, which means that its contents are not typechecked by the SDK.
Support for annotation types will be added in a subsequent PR.

cyriaquebrousse committed:
feat: add sdk for annotations api

cognite-bulldozer[bot] committed:
Merge master into CXT-488

cyriaquebrousse committed:
test: add tests for annotations verbs create, update, retrieve

cyriaquebrousse committed:
chore: revert changes in yarn.lock

andersfylling commented on packages/playground/src/types.ts:
```diff
@@ -537,3 +543,68 @@ export interface FunctionSchedulesFilter {
 }
+
+/* ************************
+   Annotations API types
```
can you put this in a types.ts file inside the annotations api folder?
Then just do the following here:
```ts
export * from './api/annotations/types';
```

cyriaquebrousse replied:
Done

andersfylling commented on packages/playground/src/types.ts:
```diff
@@ -537,3 +543,68 @@ export interface FunctionSchedulesFilter {
+  annotatedResourceIds: IdEither[];
+  annotationType?: AnnotationType;
+  creatingApp?: string;
+  creatingUser?: string | null;
```
what does null do here?

peetcremer1 replied:
`null` is a special value that signals that the annotation was not created by a particular user, i.e. the annotation was created by a service.

peetcremer1 commented on packages/playground/src/types.ts:
```diff
@@ -537,3 +543,68 @@ export interface FunctionSchedulesFilter {
+export interface AnnotationChangeById extends InternalId, AnnotationUpdate {}
+
+export interface AnnotationUpdate {
+  update: {
```
`status` seems to be missing here

cyriaquebrousse replied:
Status cannot be updated, according to the service contracts, see https://docs.cognite.com/api/playground/#operation/annotationsUpdate. Does this need to be updated?

peetcremer1 replied:
I think it needs to be updated. How else would you approve / reject an annotation before the feedback loop is implemented? It is also allowed by the API currently, see here:
https://github.com/cognitedata/context-api/blob/master/src/api/annotations_v2/schemas.py#L161

cyriaquebrousse replied:
https://github.com/cognitedata/service-contracts/pull/1583

peetcremer1 commented on packages/playground/src/types.ts:
```diff
@@ -537,3 +543,68 @@ export interface FunctionSchedulesFilter {
+************************ */
+
+export type AnnotatedResourceType = 'file';
+export type AnnotationType =
```
Perhaps it is better to not have this as a type for now and just let the user input a String, as we have currently no good way to fill in the possible values automatically.
Then, when we have a proper integration with the `annotation-types` repo, we can create this type/enum automatically and it has all the available values.

cyriaquebrousse replied:
Agreed; changed.

cyriaquebrousse committed:
chore: move annotation types to their own file

cyriaquebrousse committed:
fix: add status to annotations update

Follows https://github.com/cognitedata/service-contracts/pull/1583

cyriaquebrousse committed:
test: add test for advanced listing of annotations

andersfylling commented on packages/playground/src/index.ts:
```diff
@@ -2,3 +2,4 @@
 
 export { default as CogniteClientPlayground } from './cogniteClientPlayground';
 export * from './types';
+export * from './api/annotations/types';
```
move this to the types.ts file, then we have all the playground type/references in one location.

cyriaquebrousse committed:
chore: move export statement from index to types

andersfylling reviewed (approved):
LGTM!

~see above comment :+1:~

cyriaquebrousse committed:
test: fix create annotation test

Fix test that was requiring exact equality of two timestamps. Compare other fields instead.

cyriaquebrousse committed:
test: add test for create annotation when service is creating

peetcremer1 reviewed (approved):
LGTM

chingyulin reviewed (approved):
LGTM!
